components/authn-service/tokens/pg/migrate.go

```diff
@@ -1,183 +1,76 @@
+
+	ctx := context.TODO()
+	tx, err := db.BeginTx(ctx, nil)
 	if err != nil {
-		return errors.Wrapf(err, "acquiring advisory lock %d", constants.AdvisoryLockID)
+		return errors.Wrap(err, "beginning transaction")
 	}
 
-	_, err = c.Exec(`
-  CREATE TABLE IF NOT EXISTS migrations (
-    num INTEGER NOT NULL,
-    descr TEXT,
-    at TIMESTAMPTZ NOT NULL
-  );
-`)
-	if err != nil {
-		return errors.Wrap(err, "creating migration table")
+	if _, err := tx.ExecContext(ctx, `CREATE TABLE IF NOT EXISTS schema_migrations
+(version BIGINT NOT NULL PRIMARY KEY, dirty BOOLEAN NOT NULL)`); err != nil {
+		return errors.Wrap(err, "create new migration table 'schema_migrations'")
 	}
 
-	i := 0
-	done := false
-	var tx *sql.Tx
-	for {
-
-		tx, err = c.Begin()
-		if err != nil {
-			break
-		}
-
-		// Within a transaction, perform a single migration.
-		var num sql.NullInt64
-		n := 0
-
-		if err = tx.QueryRow(`SELECT MAX(num) FROM migrations;`).Scan(&num); err != nil {
-			err = errors.Wrap(err, "select max migration")
-			break
-		}
-		if num.Valid { // if not, n will still be 0
-			n = int(num.Int64)
-		}
-
-		if n >= len(migrations) {
-			done = true
-		} else {
-			migrationNum := n + 1
-			m := migrations[n]
-			if _, err = tx.Exec(m.statement); err != nil {
-				err = errors.Wrapf(err, "migration '%s' (%d) failed", m.descr, migrationNum)
-				break
-			}
-
-			q := `INSERT INTO migrations (num, descr, at) VALUES ($1, $2, now());`
-			if _, err = tx.Exec(q, migrationNum, m.descr); err != nil {
-				err = errors.Wrap(err, "update migration table")
-				break
-			}
-		}
-		// done with one migration, commit transaction
-		if err = tx.Commit(); err != nil {
-			break
-		}
+	// Note(sr) This needs to happen in a function because the "migrations" table might
+	// have already been dropped (i.e. on all but the first run), and such checks can only
+	// happen in functions. Just letting it crash is not an option since it aborts our
+	// transaction.
+	if _, err := tx.ExecContext(ctx, `CREATE OR REPLACE FUNCTION move_if_exists() RETURNS void AS
+$$
+BEGIN
+  IF EXISTS(SELECT * FROM information_schema.tables
+              WHERE table_schema = current_schema()
+              AND table_name = 'migrations') THEN
+     INSERT INTO schema_migrations (
+       SELECT MAX(num) AS version, FALSE AS dirty FROM migrations
+     );
+  END IF;
+END;
+$$
+LANGUAGE plpgsql;`); err != nil {
+		return errors.Wrap(err, "create move_if_exists function")
+	}
+	if _, err := tx.ExecContext(ctx, "SELECT move_if_exists()"); err != nil {
+		return errors.Wrap(err, "converting migration table columns")
+	}
+	if _, err := tx.ExecContext(ctx, "DROP TABLE IF EXISTS migrations"); err != nil {
+		return errors.Wrap(err, "dropping migration table")
+	}
 
-		if done {
-			break
-		}
-		i++ // rinse and repeat for next migration
+	if err := tx.Commit(); err != nil {
+		return errors.Wrap(err, "commit migration table migration")
 	}
 
+	m, err := migrate.New("file://"+path, url)
```

Comment:
This is done in 6 other places, all of which define their own copy of `addScheme` and use that, so might be time to add `addScheme` as a library helper function...

Reply:
....or just do the simple thing in the other places. Having looked at the `addScheme` helper after some time, it seems a bit over-engineered. 😉

Reply:
I've decided to do a cleanup sweep instead: https://github.com/chef/automate/compare/2ac97586b5bb93f26ab65acf78990e6c12479896..sr/a2-863/add-db_id-to-tokens-table